File: intervals/meeting_rooms_two.js
var minMeetingRooms = function (intervals) {
  let rooms = 0;
  let end = 0;

  // Getting the start and end times sorted
  const starts = intervals.map((a) => a[0]).sort((a, b) => a - b);
  const ends = intervals.map((a) => a[1]).sort((a, b) => a - b);

  for (let i = 0; i < intervals.length; i++) {
    // If there is a meeting that has ended by the time
    // the meeting at `start_pointer` starts
    if (ends[end] > starts[i]) rooms++;
    // Otherwise, we have an overlap
    else end++;
  }

  return rooms;
};

File: intervals/meeting_rooms_two.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        starts, ends = [], []

        for i in range(len(intervals)):
            starts.append(intervals[i][0])
            ends.append(intervals[i][1])

        starts.sort()
        ends.sort()

        # Min. number of rooms and end points
        # to the first available room
        rooms, end = 0, 0

        for i in range(len(intervals)):
            # Gives us an available room
            if ends[end] > starts[i]:
                rooms += 1
            # Check another time
            else:
                end += 1

        return rooms

# Time Complexity: O(nlogn)
# Space Complexity: O(n)